### PR TITLE
Initial InstrumenterTests

### PR DESCRIPTION
**Pull request by prdoyle**
This is an approach to unit testing that I found much easier during early POC experimentation, and I think it could be useful going forward, but I wanted to see what the team thought of it.

### The problem this solves
If a mistake is made when injecting bytecode into core JCL methods in the context of a java agent, the resulting errors are very hard to recognize and troubleshoot. The agent typically crashes, sometimes with a low-level stack trace at the C++ level. Breakpoints and single-stepping are difficult or impossible.

### The solution
Test the bytecode injection using an ordinary JUnit test. Any errors that happen get normal Java error messages, including decently informative verifier errors etc.

### How it works
The test configures `Instrumenter` to instrument our own class `ClassToInstrument` with calls to stub permission checks that are under the control of the test. It loads the instrumented class, then calls its methods, first with the checks off (to ensure the normal method logic runs correctly) and then with the checks on (to make sure the exception is thrown).

### CODE REVIEW

File: distribution/tools/entitlement-agent/impl/src/test/java/org/elasticsearch/entitlement/instrumentation/impl/InstrumenterTests.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    public static class TestEntitlementManager implements EntitlementChecks {
+        /**
+         * This allows us to test that the instrumentation is correct in both cases:
+         * if the check throws, and if it doesn't.
+         */
+        volatile boolean isActive;
+
+        @Override
+        public void checkSystemExit(Class<?> callerClass, int status) {
+            throwIfActive();
```
Comment (rjernst) [inline distribution/tools/entitlement-agent/impl/src/test/java/org/elasticsearch/entitlement/instrumentation/impl/InstrumenterTests.java:81]: could we validate the callerClass? And also the status (to make sure it is passed through correctly)?

**Review by rjernst — APPROVED** [review]
Nice, good idea. I have one small comment on additional assertions. I also think we should expand a bit (will need to test non-static methods too), but that can happen in followups.

**Review by ldematte — APPROVED** [review]
I like it! I will probably be using this very soon 👍 
LGTM, but it would be great if you could add the asserts Ryan suggested

**Comment by ldematte** [comment]
(IDK why github shows failures - if you jump to buildkite, everything is green)

File: distribution/tools/entitlement-agent/impl/src/test/java/org/elasticsearch/entitlement/instrumentation/impl/InstrumenterTests.java (hunk 1 above)
Comment (prdoyle) [reply distribution/tools/entitlement-agent/impl/src/test/java/org/elasticsearch/entitlement/instrumentation/impl/InstrumenterTests.java:81]: Great idea!

**Comment by prdoyle** [comment]
[This failure](https://gradle-enterprise.elastic.co/s/b4sn7q7ftwfom/tests/task/:build-tools-internal:integTest/details/org.elasticsearch.gradle.internal.PublishPluginFuncTest/hides%20runtime%20dependencies%20and%20handles%20shadow%20dependencies?top-execution=1) seems to be a difference in the reported software licenses. There's no way I caused that.

Let me merge main and see if that helps.

**Comment by prdoyle** [comment]
Note to self: the "pre-activation" tests for void methods will pass even if the target method isn't called at all. I wonder if we'd be better off with a Mockito mock where we can assert that the target method got called. 🤔

**Comment by craigtaverner** [comment]
> The failures seem to be related to #114625

The tests are now muted, so things should be fine now, even before the above mentioned fix is merged.

**Pull request merged**